### PR TITLE
[SystemC] Lower Interop pass for verilated operation

### PR DESCRIPTION
**Pull request by maerhart**
A first version for lowering the `systemc.interop.verilated` operation.

### CODE REVIEW

File: lib/Dialect/SystemC/Transforms/SystemCLowerInstanceInterop.cpp [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+
+  LogicalResult
+  matchAndRewrite(InteropVerilatedOp op, OpAdaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const override {
+    // TODO: instead of hardcoding the verilated module's class name, it should
+    // be derived from a configs attribute as this can be specified via the CLI
+    // arguments of verilator
+    // stateType ::= VModuleName*
+    std::string tn = "V";
+    tn += op.getModuleName();
```
Comment (seldridge) [inline lib/Dialect/SystemC/Transforms/SystemCLowerInstanceInterop.cpp:43]: Nit: consider using `SmallString`.

File: lib/Dialect/SystemC/Transforms/SystemCLowerInstanceInterop.cpp [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+          loc, inputValues[i].getType(), state,
+          inputNames[i].cast<StringAttr>(), MemberAccessKind::Arrow);
+      updateBuilder.create<AssignOp>(loc, member,
+                                     updateOp.getBody()->getArgument(i + 1));
+    }
+
+    // Call 'eval'.
+    auto evalFunc = updateBuilder.create<MemberAccessOp>(
+        loc, FunctionType::get(updateBuilder.getContext(), {}, {}), state,
+        "eval", MemberAccessKind::Arrow);
```
Comment (seldridge) [inline lib/Dialect/SystemC/Transforms/SystemCLowerInstanceInterop.cpp:115]: Raw strings always seem weird. However, this "eval" is only ever used once. It is likely fine to not hoist this into a header.

File: test/Dialect/SystemC/Transforms/interop.mlir [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+  // CHECK-NEXT: }
+  // CHECK-NEXT: interop.procedural.dealloc cpp [[STATE2]] : !emitc.ptr<!emitc.opaque<"VEmpty">> {
+  // CHECK-NEXT: ^bb0(%arg0: !emitc.ptr<!emitc.opaque<"VEmpty">>):
+  // CHECK-NEXT:   systemc.cpp.delete %arg0 : !emitc.ptr<!emitc.opaque<"VEmpty">>
+  // CHECK-NEXT: }
+  systemc.interop.verilated "inst1" @Empty () -> ()
+
+  // CHECK-NEXT: hw.output {{%.+}} : i32
+  hw.output %c : i32
+}
```
Comment (seldridge) [inline test/Dialect/SystemC/Transforms/interop.mlir:49]: I can't think of a good way to golf down this test other than decreasing the arity of `@Bar` to one input port. That doesn't save much, though. 😅

**Review by seldridge — APPROVED** [review]
Looks good. I'm happy to see this whole thing coming together after a ton of infra work. 👍

File: lib/Dialect/SystemC/Transforms/SystemCLowerInstanceInterop.cpp (hunk 1 above)
Comment (maerhart) [reply lib/Dialect/SystemC/Transforms/SystemCLowerInstanceInterop.cpp]: Sorry that slipped through.

**Pull request merged**